README.md
# heissepreise-influx

<!---
[![start with why](https://img.shields.io/badge/start%20with-why%3F-brightgreen.svg?style=flat)](http://www.ted.com/talks/simon_sinek_how_great_leaders_inspire_action)
--->
[![GitHub release](https://img.shields.io/github/release/elbosso/heissepreise-influx/all.svg?maxAge=1)](https://GitHub.com/elbosso/heissepreise-influx/releases/)
[![GitHub tag](https://img.shields.io/github/tag/elbosso/heissepreise-influx.svg)](https://GitHub.com/elbosso/ImageTagManager/tags/)
[![GitHub license](https://img.shields.io/github/license/elbosso/heissepreise-influx.svg)](https://github.com/elbosso/ImageTagManager/blob/master/LICENSE)
[![GitHub issues](https://img.shields.io/github/issues/elbosso/heissepreise-influx.svg)](https://GitHub.com/elbosso/ImageTagManager/issues/)
[![GitHub issues-closed](https://img.shields.io/github/issues-closed/elbosso/heissepreise-influx.svg)](https://GitHub.com/elbosso/ImageTagManager/issues?q=is%3Aissue+is%3Aclosed)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](https://github.com/elbosso/ImageTagManager/issues)
[![GitHub contributors](https://img.shields.io/github/contributors/elbosso/heissepreise-influx.svg)](https://GitHub.com/elbosso/ImageTagManager/graphs/contributors/)
[![Github All Releases](https://img.shields.io/github/downloads/elbosso/heissepreise-influx/total.svg)](https://github.com/elbosso/ImageTagManager)
[![Website elbosso.github.io](https://img.shields.io/website-up-down-green-red/https/elbosso.github.io.svg)]([https://elbosso.github.io/](https://elbosso.github.io/supermarktpreisentwicklung_influxdb.html#content))

## Überblick

Dieses Projekt importiert historische Preisdaten einiger deutscher/österreichischer Supermärkte in eine InfluxDB 
Zeitreihendatenbank und bietet Mittel, diese durch aktuelle Preisinformationen fortzuschreiben.

```
mvn compile package
```

and then starting the resulting monolithic jar file by issuing

```
$JAVA_HOME/bin/java -jar target/heissepreise-influx-<version>-jar-with-dependencies.jar
```
## Hintergrund

[Supermarktpreisentwicklung in InfluxDB](https://elbosso.github.io/supermarktpreisentwicklung_influxdb.html#content)

## Benutzung

Die Anwendung selbst verfügt über zwei Subcommands: `Update` und `Import`. Update funktioniert derzeit nur für Reqe (DE)
Zu beachten ist, dass die Parameter `INFLUXDB_DATA_MAX_VALUES_PER_TAG` und `INFLUXDB_DATA_MAX_SERIES_PER_DATABASE`
der InfluxDB, die die Daten aufnehmen soll auf *0* gesetzt werden müssen (https://docs.influxdata.com/influxdb/v1/administration/config/#data-settings).

Der Import kann beispielsweise mittels des Kommandos

```
$JAVA_HOME/bin/java -jar target/heissepreise-influx-<version>-jar-with-dependencies.jar Import -l INFO -d prices -H influxdb.docker.lab -P 443 -S HTTPS /home/elbosso/temp/heisse-preise.json
```
gestartet werden - wenn vorher die historischen Daten [heruntergeladen](https://github.com/badlogic/heissepreise) wurden.

### Import

```
Usage: HeissePreiseInfluxdbInterface Import [-hV] [-a=<measurementName>]
       -d=<influxDbDatabasename> [-H=<influxDbHost>] [-l=<logLevel>]
       [-p=<influxDbPassword>] [-P=<influxDbPort>] [-S=<influxDbScheme>]
       [-u=<influxDbUsername>] <jsonFile>
Importiert historische Preisdaten von Supermärkten.
      <jsonFile>   Diese Datei enthält die historischen Daten der Supermärkte.
                     Verfügbar zum Beispiel hier: https://heisse-preise.
                     io/data/latest-canonical.json *(Achtung: über 100 MB!).*
                     Das Ursprungsprojekt ist https://github.
                     com/badlogic/heissepreise .
  -a, -measurementName-=<measurementName>
                   measurementName (default: Product). Das ist der Name des
                     Measurement, unter dem die Daten importiert werden.
  -d, --influxDbDatabasename=<influxDbDatabasename>
                   influxDbDatabasename. Der Name der Influx 1.x Datenbank, in
                     die die Daten geschrieben werden sollen.
  -h, --help       Show this help message and exit.
  -H, --influxDbHost=<influxDbHost>
                   influxDbHost (default: localhost). Der Name oder die
                     IP-Adresse zur Verbindung mit der Influx 1.x Datenbank.
  -l, --logLevel=<logLevel>
                   logLevel (default: WARN). Der Loglevel bestimmt die
                     Ausführlichkeit der Ausgaben während der Bearbeitung.
                     Erleubte Werte sind TRACE, DEBUG, INFO, WARN und ERROR.
  -p, --influxDbPassword=<influxDbPassword>
                   influxDbPassword (default: null). Das PAsswort für die
                     Authentifizierung an der Influx 1.x Datenbank - falls
                     benötigt.
  -P, --influxDbPort=<influxDbPort>
                   influxDbPort (default: 8086). Der Port zur Verbindung mit
                     der Influx 1.x Datenbank
  -S, --influxDbScheme=<influxDbScheme>
                   influxDbScheme (default: HTTP). Erlaubte Werte sind HTTP für
                     Verbindungen zur Influx1.x Datenbank ohne TLS und HTTPS
                     für TLS-verschlüsselte Verbindungen.
  -u, --influxDbUsername=<influxDbUsername>
                   influxDbUsername (default: null). Der Nutzername für die
                     Authentifizierung an der Influx 1.x Datenbank - falls
                     benötigt.
  -V, --version    Print version information and exit.
```

### Update

```
Usage: HeissePreiseInfluxdbInterface Update [-hV] [-a=<measurementName>]
       -d=<influxDbDatabasename> [-H=<influxDbHost>] [-l=<logLevel>]
       [-p=<influxDbPassword>] [-P=<influxDbPort>] [-s=<storeNumber>]
       [-S=<influxDbScheme>] [-u=<influxDbUsername>]
Aktualisiert Preisdaten von Rewe Supermärkten. Derzeit wird nur Rewe (DE)
unterstützt. Weiterewerden folgen. Das Update erfolgt so, dass er mit dem
Import historischer Daten (subcommand Update) zusammen funktioniert. Alle
Informationen werden in ein Measurement importiert und die Informationen zu
einzelnen Produkten können anhang des Tags "name" differenziert werden.
  -a, -measurementName-=<measurementName>
                  measurementName (default: Product). Das ist der Name des
                    Measurement, unter dem die Daten importiert werden.
  -d, --influxDbDatabasename=<influxDbDatabasename>
                  influxDbDatabasename. Der Name der Influx 1.x Datenbank, in
                    die die Daten geschrieben werden sollen.
  -h, --help      Show this help message and exit.
  -H, --influxDbHost=<influxDbHost>
                  influxDbHost (default: localhost). Der Name oder die
                    IP-Adresse zur Verbindung mit der Influx 1.x Datenbank.
  -l, --logLevel=<logLevel>
                  logLevel (default: WARN). Der Loglevel bestimmt die
                    Ausführlichkeit der Ausgaben während der Bearbeitung.
                    Erleubte Werte sind TRACE, DEBUG, INFO, WARN und ERROR.
  -p, --influxDbPassword=<influxDbPassword>
                  influxDbPassword (default: null). Das PAsswort für die
                    Authentifizierung an der Influx 1.x Datenbank - falls
                    benötigt.
  -P, --influxDbPort=<influxDbPort>
                  influxDbPort (default: 8086). Der Port zur Verbindung mit der
                    Influx 1.x Datenbank
  -s, --storeNumber=<storeNumber>
                  storeNumber (default: 440405). Das ist die Nummer, die für
                    die Abfrage dere Preisdaten aus der ReweDigital-API
                    benötigt wird. Die Preise der Produkte können nur
                    Store-spezifisch abgefragt werden, da das Angebot sich
                    zwischen den einzelnen Stores unterscheidet. Möchte man
                    einen anderen als den als default hinterlegten Store
                    abfragen, kann man die benötigte Nummer über einen anderen
                    Endpunkt der API herausfinden (https://github.
                    com/foo-git/rewe-discounts/blob/master/rewe_discounts/rewe_d
                    iscounts.py).
  -S, --influxDbScheme=<influxDbScheme>
                  influxDbScheme (default: HTTP). Erlaubte Werte sind HTTP für
                    Verbindungen zur Influx1.x Datenbank ohne TLS und HTTPS für
                    TLS-verschlüsselte Verbindungen.
  -u, --influxDbUsername=<influxDbUsername>
                  influxDbUsername (default: null). Der Nutzername für die
                    Authentifizierung an der Influx 1.x Datenbank - falls
                    benötigt.
  -V, --version   Print version information and exit.
```


CHANGELOG.md
          # Changelog

Changelog of this project.


**Update README.md**

 * added documentation

[85cc5df858a5625](https://github.com/elbosso/heissepreise-influx/commit/85cc5df858a5625) Jürgen Key *2023-09-27 15:01:50*

**corrected meta information and made a start on documentation**


[2e2fe0754b9a88c](https://github.com/elbosso/heissepreise-influx/commit/2e2fe0754b9a88c) Juergen Key *2023-09-27 11:44:53*




src/main/java/de/elbosso/tools/heissepreiseinflux/Import.java
package de.elbosso.tools.heissepreiseinflux;

import de.elbosso.util.io.LineProtocolWriter;
import de.netsysit.ui.dialog.LoginDialog;
import org.json.JSONException;
import org.slf4j.event.Level;
import picocli.CommandLine;

import java.io.IOException;
import java.text.ParseException;

@CommandLine.Command(name = "Import", mixinStandardHelpOptions = true, version = "HeissePreiseInfluxdbInterface 0.1.0",
        description = "Importiert historische Preisdaten von Supermärkten.")
public class Import extends java.lang.Object implements Runnable
{
    private final static org.slf4j.Logger CLASS_LOGGER =org.slf4j.LoggerFactory.getLogger(Import.class);
    private final static org.slf4j.Logger EXCEPTION_LOGGER =org.slf4j.LoggerFactory.getLogger("ExceptionCatcher");
    @CommandLine.Option(names = {"-S", "--influxDbScheme"}, description = "influxDbScheme (default: ${DEFAULT-VALUE})", defaultValue = "HTTP")
    private HeissePreiseInfluxdbInterface.Scheme influxDbScheme;
    @CommandLine.Option(names = {"-H", "--influxDbHost"}, description = "influxDbHost (default: ${DEFAULT-VALUE})", defaultValue = "localhost")
    private java.lang.String influxDbHost;
    @CommandLine.Option(names = {"-P", "--influxDbPort"}, description = "influxDbPort (default: ${DEFAULT-VALUE})", defaultValue = "8086")
    private int influxDbPort;
    @CommandLine.Option(names = {"-u", "--influxDbUsername"}, description = "influxDbUsername (default: ${DEFAULT-VALUE})", defaultValue = CommandLine.Option.NULL_VALUE)
    private java.lang.String influxDbUsername;
    @CommandLine.Option(names = {"-p", "--influxDbPassword"}, description = "influxDbPassword (default: ${DEFAULT-VALUE})", defaultValue = CommandLine.Option.NULL_VALUE)
    private java.lang.String influxDbPassword;
    @CommandLine.Option(names = {"-d", "--influxDbDatabasename"}, description = "influxDbDatabasename", required = true)
    private java.lang.String influxDbDatabasename;
    //    @CommandLine.Option(names = {"-m", "--mode"}, description = "mode (default: ${DEFAULT-VALUE})", defaultValue = "initialImport")
//    private Mode mode;
    @CommandLine.Option(names = {"-l", "--logLevel"}, description = "logLevel (default: ${DEFAULT-VALUE})", defaultValue = "WARN")
    private Level logLevel;
    @CommandLine.Option(names = {"-a", "-measurementName-"}, description = "measurementName (default: ${DEFAULT-VALUE})", defaultValue = "Product")
    private java.lang.String measurementName;
    @CommandLine.Parameters(index = "0", description = "File containing the historic data in JSON format")
    private java.io.File jsonFile;

    private static java.text.DateFormat df=new java.text.SimpleDateFormat("yyyy-MM-dd");
    private de.elbosso.util.io.InfluxDBHttpInterface influxDBHttpInterface;
    private LineProtocolWriter lineProtocolWriter;
    private long counter;
    @Override
    public void run()
    {
        de.elbosso.util.Utilities.configureBasicStdoutLogging(logLevel);
        CLASS_LOGGER.trace("starting up");
        try
        {
            doInitialImport();
        } catch (Throwable t)
        {
            EXCEPTION_LOGGER.error(t.getMessage(), t);
        }
    }

    private void handle(org.json.JSONObject product) throws JSONException, ParseException, IOException
    {
        org.json.JSONArray pHistory = product.getJSONArray("priceHistory");
        String lineFormat = null;
        for (int i = 0; i < pHistory.length(); ++i)
        {
            org.json.JSONObject price = pHistory.getJSONObject(i);
            java.util.Map<String, String> tags = new java.util.HashMap();
            java.util.Map<String, Object> fields = new java.util.HashMap();
            tags.put("store", product.getString("store"));
            tags.put("name", product.getString("name"));
            tags.put("id", product.getString("id"));
            if (product.has("unit"))
                tags.put("unit", product.getString("unit"));
            try
            {
                if (product.has("quantity"))
                    fields.put("amount", product.getInt("quantity"));
                fields.put("currentPrice", price.getDouble("price"));
                String timestampS = price.getString("date");
                java.util.Date timestamp = df.parse(timestampS);
                lineFormat = lineProtocolWriter.write(timestamp, tags, fields);
                CLASS_LOGGER.debug(lineFormat);
                influxDBHttpInterface.send(lineFormat);
                ++counter;
                if (counter % 100000 == 0)
                    CLASS_LOGGER.info("processed item {}", counter);
            } catch (Throwable t)
            {
                EXCEPTION_LOGGER.warn(t.getMessage() + "\n" + product.toString(2));
            }
        }
    }

    private void doInitialImport() throws JSONException, IOException, ParseException
    {
        CLASS_LOGGER.trace("doInitialImport");
        lineProtocolWriter = new LineProtocolWriter(measurementName);
        LoginDialog.BasicCredentials bc = null;
        if ((influxDbUsername != null) && (influxDbPassword != null))
            bc = new LoginDialog.BasicCredentials(influxDbUsername, influxDbPassword);
        influxDBHttpInterface = new de.elbosso.util.io.InfluxDBHttpInterface(influxDbScheme.name(), influxDbHost, influxDbPort, influxDbDatabasename, bc);
        java.io.FileInputStream fis = new java.io.FileInputStream(jsonFile);
        java.io.InputStreamReader isr = new java.io.InputStreamReader(fis);
        java.io.BufferedReader br = new java.io.BufferedReader(isr);
        org.json.JSONTokener tok = new org.json.JSONTokener(br);
        org.json.JSONArray history = new org.json.JSONArray(tok);
        java.util.List<String> lineFormats = new java.util.LinkedList();
        for (int i = 0; i < history.length(); ++i)
        {
            org.json.JSONObject obj = history.getJSONObject(i);
            if (obj.getString("store").equalsIgnoreCase("penny") == false)
            {
                if (obj.has("priceHistory"))
                {
                    CLASS_LOGGER.debug("{} {}", i, obj.getString("name"));
                    handle(obj);
                }
            }
        }
        influxDBHttpInterface.flush();
    }

}


src/main/java/de/elbosso/tools/heissepreiseinflux/HeissePreiseInfluxdbInterface.java
package de.elbosso.tools.heissepreiseinflux;

import picocli.CommandLine;

@CommandLine.Command(name = "HeissePreiseInfluxdbInterface", mixinStandardHelpOptions = true, version = "HeissePreiseInfluxdbInterface 0.1.0",
        description = "Importiert historische Preisdaten von Supermärkten und aktualisiert diese mit aktuellen Informationen."
,subcommands = {Import.class, Update.class})
public class HeissePreiseInfluxdbInterface extends java.lang.Object
{
	private final static org.slf4j.Logger CLASS_LOGGER =org.slf4j.LoggerFactory.getLogger(HeissePreiseInfluxdbInterface.class);
	private final static org.slf4j.Logger EXCEPTION_LOGGER =org.slf4j.LoggerFactory.getLogger("ExceptionCatcher");

    private HeissePreiseInfluxdbInterface()
    {
        super();
    }


    public static void main(java.lang.String[] args)
    {
        CommandLine commandLine=new CommandLine(new HeissePreiseInfluxdbInterface());
        commandLine.execute(args);
    }
/*    enum Mode
    {
        initialImport,
        updateRewe;
    }
*/    enum Scheme
    {
        HTTP,
        HTTPS;
    }

}


src/main/java/de/elbosso/tools/heissepreiseinflux/Update.java
package de.elbosso.tools.heissepreiseinflux;

import de.elbosso.util.io.LineProtocolWriter;
import de.elbosso.util.lowlevel.ProcessCapture;
import de.netsysit.ui.dialog.LoginDialog;
import org.json.JSONException;
import org.slf4j.event.Level;
import picocli.CommandLine;

import java.io.IOException;
import java.net.MalformedURLException;
import java.text.ParseException;
import java.util.Locale;

@CommandLine.Command(name = "Update", mixinStandardHelpOptions = true, version = "HeissePreiseInfluxdbInterface 0.1.0",
        description = "Aktualisiert Preisdaten von Supermärkten.")
public class Update extends java.lang.Object implements java.lang.Runnable
{
    private final static org.slf4j.Logger CLASS_LOGGER =org.slf4j.LoggerFactory.getLogger(Update.class);
    private final static org.slf4j.Logger EXCEPTION_LOGGER =org.slf4j.LoggerFactory.getLogger("ExceptionCatcher");
    @CommandLine.Option(names = {"-s", "--storeNumber"}, description = "storeNumber (default: ${DEFAULT-VALUE})", defaultValue = "440405")
    private String storeNumber;
    @CommandLine.Option(names = {"-S", "--influxDbScheme"}, description = "influxDbScheme (default: ${DEFAULT-VALUE})", defaultValue = "HTTP")
    private HeissePreiseInfluxdbInterface.Scheme influxDbScheme;
    @CommandLine.Option(names = {"-H", "--influxDbHost"}, description = "influxDbHost (default: ${DEFAULT-VALUE})", defaultValue = "localhost")
    private java.lang.String influxDbHost;
    @CommandLine.Option(names = {"-P", "--influxDbPort"}, description = "influxDbPort (default: ${DEFAULT-VALUE})", defaultValue = "8086")
    private int influxDbPort;
    @CommandLine.Option(names = {"-u", "--influxDbUsername"}, description = "influxDbUsername (default: ${DEFAULT-VALUE})", defaultValue = CommandLine.Option.NULL_VALUE)
    private java.lang.String influxDbUsername;
    @CommandLine.Option(names = {"-p", "--influxDbPassword"}, description = "influxDbPassword (default: ${DEFAULT-VALUE})", defaultValue = CommandLine.Option.NULL_VALUE)
    private java.lang.String influxDbPassword;
    @CommandLine.Option(names = {"-d", "--influxDbDatabasename"}, description = "influxDbDatabasename", required = true)
    private java.lang.String influxDbDatabasename;
    //    @CommandLine.Option(names = {"-m", "--mode"}, description = "mode (default: ${DEFAULT-VALUE})", defaultValue = "initialImport")
//    private Mode mode;
    @CommandLine.Option(names = {"-l", "--logLevel"}, description = "logLevel (default: ${DEFAULT-VALUE})", defaultValue = "WARN")
    private Level logLevel;
    @CommandLine.Option(names = {"-a", "-measurementName-"}, description = "measurementName (default: ${DEFAULT-VALUE})", defaultValue = "Product")
    private java.lang.String measurementName;

    private de.elbosso.util.io.InfluxDBHttpInterface influxDBHttpInterface;
    private LineProtocolWriter lineProtocolWriter;
    private long counter;
    private static java.text.NumberFormat nf = java.text.NumberFormat.getNumberInstance(Locale.GERMAN);
    private static java.util.regex.Pattern fullPattern = java.util.regex.Pattern.compile("(\\d+)\\s*?(\\S+)\\s*?.*?\\((.*?)=(.*?)\\)");
    private static java.util.regex.Pattern pattern = java.util.regex.Pattern.compile("(\\d+)\\s*?(\\S+)\\s*?.*?");

    public void run()
    {
        de.elbosso.util.Utilities.configureBasicStdoutLogging(logLevel);
        CLASS_LOGGER.trace("starting up");
        try
        {
            doUpdate();
        } catch (Throwable t)
        {
            EXCEPTION_LOGGER.error(t.getMessage(), t);
        }
    }
    private java.net.URL makeUrl(int pageNumber,java.lang.String marketCode) throws MalformedURLException
    {
        java.lang.String serviceTypes="PICKUP";
        java.lang.String searchTerm="*";
        int perPage=250;
        java.lang.String endpoint="https://mobile-api.rewe.de/api/v3/product-search?searchTerm={0}&page={1}&sorting=RELEVANCE_DESC&objectsPerPage={2}&marketCode={3}&serviceTypes={4}";
        java.lang.String request=java.text.MessageFormat.format(endpoint,searchTerm,pageNumber,perPage,marketCode,serviceTypes);
        CLASS_LOGGER.debug(request);
        return new java.net.URL(request);
    }
    private java.lang.String getPageUsingCurl(int pageNumber) throws IOException, InterruptedException
    {
        java.lang.String marketCode=storeNumber;
        java.lang.String serviceTypes="PICKUP";
        java.lang.String[] commands=new java.lang.String[]{
                "curl",
                makeUrl(pageNumber,marketCode).toString(),
                "-H",
                "Rd-Service-Types: "+serviceTypes,
                "-H",
                "Rd-Market-Id: "+marketCode,
        };
        ProcessCapture processCapture=new ProcessCapture(commands);
        java.lang.Thread t=new java.lang.Thread(processCapture);
        t.start();
        t.join();
        java.lang.String jsonAsString=processCapture.getStdout();
        return jsonAsString;
    }
    private void handle(org.json.JSONObject json, java.lang.String store) throws JSONException, ParseException, IOException
    {
        org.json.JSONArray products = json.getJSONArray("products");
        CLASS_LOGGER.debug("products {}", products.length());
        java.util.Map<String, String> tags = null;
        java.util.Map<String, Object> fields = null;
        for (int i = 0; i < products.length(); ++i)
        {
            tags = new java.util.HashMap();
            fields = new java.util.HashMap();
            org.json.JSONObject product = products.getJSONObject(i);
            tags.put("store", store);
            tags.put("hasBioCide", java.lang.Boolean.toString(product.getBoolean("hasBioCide")));
            tags.put("name", product.getString("name"));
            tags.put("id", product.getString("id"));
            tags.put("listingId", product.getString("listingId"));
            if (product.has("tags"))
            {
                org.json.JSONArray jtags = product.getJSONArray("tags");
                for (int j = 0; j < jtags.length(); ++j)
                {
                    tags.put(jtags.getString(j), "true");
                }
            }
            java.lang.String currentPrice = product.getString("currentPrice");
            currentPrice = currentPrice.substring(0, currentPrice.indexOf(' '));
            CLASS_LOGGER.trace("{} {}", currentPrice, nf.parse(currentPrice));
            fields.put("currentPrice", nf.parse(currentPrice));
            if(product.has("grammage"))
            {
                java.lang.String input = product.getString("grammage");
                java.util.regex.Matcher matcher = fullPattern.matcher(input);
                if (matcher.matches())
                {
                    CLASS_LOGGER.trace(input);
                    tags.put("amount", matcher.group(1));
                    tags.put("unit", matcher.group(2));
                    tags.put("relativeAmount", matcher.group(3));
                    java.lang.String relativePrice = matcher.group(4).trim();
                    relativePrice = relativePrice.substring(0, relativePrice.indexOf(' '));
                    CLASS_LOGGER.trace("{} {}", relativePrice, nf.parse(relativePrice));
                    fields.put("relativePrice", nf.parse(relativePrice));
                }
                else
                {
                    matcher = pattern.matcher(input);
                    if (matcher.matches())
                    {
                        CLASS_LOGGER.trace(input);
                        fields.put("amount", java.lang.Integer.parseInt(matcher.group(1)));
                        tags.put("unit", matcher.group(2));
                    }
                }
            }
            java.lang.String lineFormat = lineProtocolWriter.write(tags,fields);
            CLASS_LOGGER.debug(lineFormat);
            influxDBHttpInterface.send(lineFormat);
            ++counter;
            if (counter % 1000 == 0)
                CLASS_LOGGER.info("processed item {}", counter);
        }
    }
    private void doUpdate() throws JSONException, IOException, ParseException, InterruptedException
    {
        CLASS_LOGGER.trace("doInitialImport");
        lineProtocolWriter = new LineProtocolWriter(measurementName);
        LoginDialog.BasicCredentials bc = null;
        if ((influxDbUsername != null) && (influxDbPassword != null))
            bc = new LoginDialog.BasicCredentials(influxDbUsername, influxDbPassword);
        influxDBHttpInterface = new de.elbosso.util.io.InfluxDBHttpInterface(influxDbScheme.name(), influxDbHost, influxDbPort, influxDbDatabasename, bc);
        int pageNumber=1;
        java.lang.String jsonAsString=getPageUsingCurl(pageNumber);
        org.json.JSONObject json=new org.json.JSONObject(jsonAsString);
        handle(json,"reweDe");
        int totalPages=json.getInt("totalPages");
        CLASS_LOGGER.debug(json.getString("page"));
        java.util.List<java.lang.String> lineFormats=new java.util.LinkedList();
        while(totalPages>pageNumber)
        {
            ++pageNumber;
            jsonAsString=getPageUsingCurl(pageNumber);
            json=new org.json.JSONObject(jsonAsString);
            CLASS_LOGGER.debug(json.getString("page"));
            handle(json,"reweDe");
            totalPages=json.getInt("totalPages");
            CLASS_LOGGER.debug(jsonAsString);
        }
        influxDBHttpInterface.flush();

    }

    enum Market
    {
        rewede;
    }
}
